stream/stream_file.c

```diff
@@ -332,6 +332,17 @@ static int open_f(stream_t *stream, const struct stream_open_args *args)
     setmode(p->fd, O_BINARY);
 #endif
 
+#if HAVE_SEEK_DATA
+    if (stream->mode == STREAM_READ) {
+        off_t first_data = lseek(p->fd, 0, SEEK_DATA);
+        if (first_data == (off_t)-1 && errno == ENXIO) {
+            MP_ERR(stream, "File is empty or all sparse (has no data).\n");
+            s_close(stream);
+            return STREAM_ERROR;
```

Comment:
Is opening empty file really a stream error? What if it is produced at the same time as mpv want to read it? Why it errors out when opening the file instead of returning eof in read?

Reply:
From `man 2 lseek`:

>
>  **`ENXIO`**  
>               whence  is `SEEK_DATA` or `SEEK_HOLE`, and `offset` is beyond the
>                end of the file, ***or  whence  is  `SEEK_DATA`  and  `offset`  is
>                within a hole at the end of the file.***
>

Not sure if that's what you're asking about, but 0 offset in an empty file is technically *within a hole at the end of the file*!

Reply:
I'm asking whether it is reasonable to error out when opening empty files. From `open_f` pov it would be unexpected to see an error if there is a valid file in my opinion. And if anything stream read can return EOF on such files.

Reply:
Right.

I guess it's a question of wether keeping a conceptual symmetry with IO open is worth it, or even helpful.

Is an empty file a valid media stream? If the answer is no. Then it's reasonable to error on open, IMHO.

For me, the error from this PR is more useful from a user perspective than  `Failed to recognize file format.` which users get on actual 0-sized empty files now.

Sparse-awareness on the reader side would be more involved. An optimization there would be using something like `calloc` for hole-bytes, skipping IO. Returning EOF if all remaining bytes are sparse would also break symmetry with IO. And in this case, it is actually practically relevant, for partially sparse files at least.

Reply:
ok